### ISSUE TITLE
VRL: add `parse_json_object` as a convenient way to parse JSON objects

### ISSUE THREAD

**Issue by JeanMertz** [vectordotdev/vector#9927]
The `parse_json` function is frequently used to parse the `.message` field into a JSON object. Technically, a string such as `"true"` is also valid JSON, in that it produces the boolean JSON value `true`. Because of this, the return type of the `parse_json` function is any valid JSON type.

This is annoying when you just want to parse the message as a JSON object, e.g.:

```coffee
. |= object!(parse_json!(.message))
```

I propose we add a convenience function `parse_json_object` that fails if the resulting type isn't an object.

This allows you to write:

```coffee
. |= parse_json_object!(.message)
```

Which isn't _much_ shorter, but it is less cluttered with verbose syntax of wrapping two functions.

I also thought of allowing something like `parse_json!(.message, kind: "object")`, but that feels like an awkward interface to have.

Note that there is also the `merge` function, which already works this way because it requires an object as input:

```coffee
. = merge!(., .message) # fails if `.` or `.message` isn't an object
```

But I've seen a sufficient number of people use `. |= ...` in the wild, that it might make sense to optimize that pattern.

I'm also open to alternatives that improve this use-case (perhaps "just" documenting this better could be a first step).

**Comment by hhromic** [comment]
> I also thought of allowing something like `parse_json!(.message, kind: "object")`, but that feels like an awkward interface to have.

I was thinking on this exact alternative (although my proposal was to use `type` as the parameter name, not `kind`, with a possible value being `any` as well -and be the default-).

This because I was thinking that if you offer `parse_json_object()`, users (me included) would also expect (for completeness) to have `parse_json_array()` and `parse_json_primitive()` (or literal) available as well.

**Comment by JeanMertz** [comment]
> This because I was thinking that if you offer `parse_json_object()`, users (me included) would also expect (for completeness) to have `parse_json_array()` and `parse_json_primitive()` (or literal) available as well.

Maybe, but I don't consider "completeness" to be an objective we should strive for. `parse_json` together with one of the type-coercion functions fulfil those needs, it's just that this need in particular is so wide-spread that it _might_ warrant a separate function.

I'm more inclined to _not_ do this (and solve the original problem with better documentation) than to expand the stdlib with `parse_json_array` etc.

**Comment by hhromic** [comment]
Makes sense. We personally process JSON objects and never had the need for arrays alone, but was just commenting in case we need and the option is there (assuming is a low hanging fruit to implement).

In any case all the enhancements you plan are very welcomed for us, guys, really nice job.

**Comment by jszwedko** [comment]
Yeah, I think this is a nice quality of life improvement since the majority of the time people want an object, in my experience.

 I can definitely understand the proposal to add other JSON parsing methods just for consistency, but feel they are unlikely to be used nearly as often. Though, `parse_json_array` might see some usage.

**Comment by hhromic** [comment]
Yes perhaps my `parse_json_primitive()` suggestion was a bit too far stretched ^_^!.
And yes, I was also thinking that the array one might get some uses somewhere.

**Comment by JeanMertz** [comment]
To be clear, you can still use `array!(parse_json(.))`, which is a bit less ergonomic, but its usage is probably not high enough that `parse_json_array` should be needed.

I'm also not entirely on board with `parse_json_object` yet, there might be a higher level solution to solve this problem.

One potential solution I was thinking about is to have the error-assign operation catch invalid merge assignments, that is:

```coffee
ok, err |= parse_json!(.message)
```

Currently, this isn't valid:

```
error[E104]: unnecessary error assignment
  ┌─ :1:5
  │
1 │ ok, err |= parse_json!(.message)
  │ --  ^^^    --------------------- because this expression can't fail
  │ │   │
  │ │   this error assignment is unnecessary
  │
  = see documentation about error handling at https://errors.vrl.dev/#handling
  = learn more about error code 104 at https://errors.vrl.dev/104
  = see language documentation at https://vrl.dev
```

But we could make it so that the error assignment catches an invalid merge-assign operation if the rhs type isn't an object at runtime.

So instead of this:

```coffee
. |= parse_json!(.message)
```

You could write:

```coffee
# do something with `err` at runtime
., err |= parse_json!(.message)
```

Or event:

```coffee
# ignore `err` at runtime
., _ |= parse_json!(.message)
```

This is one potential solution that came to mind, there might be others as well.

**Comment by JeanMertz** [comment]
Ah, actually, the above wouldn't work because the error-assign operation assigns the "empty" value to the `ok` target if the rhs expression fails.

So in this case, after failing a parse, if `parse_json` where to result in a boolean, then `.` would be assigned the `false` value, which isn't allowed to be assigned to `.`.

So yeah, you'd have to do it in two steps:

```coffee
value, _ |= parse_json!(.message)
. = object!(value)
```

Which wins you exactly... nothing 😅

I wonder if we should prevent the err-assign operation to assign a value to the `ok` target _if_ that target already exists. It would be a breaking change, and it might be confusing to people now that I think about it, but it would at least solve the above issue.

**Comment by hhromic** [comment]
> Currently, this isn't valid:
> 
> ```
> error[E104]: unnecessary error assignment
>   ┌─ :1:5
>   │
> 1 │ ok, err |= parse_json!(.message)
>   │     ^^^    --------------------- because this expression can't fail
>   │     │
>   │     this error assignment is unnecessary
>   │
>   = see documentation about error handling at https://errors.vrl.dev/#handling
>   = learn more about error code 104 at https://errors.vrl.dev/104
>   = see language documentation at https://vrl.dev
> ```

I think you meant this?
```coffee
ok, err = parse_json(.message)
# "function call error for \"parse_json\" at (10:30): expected \"string\", got \"null\""
```

You were using the `!` modifier, thus indeed making the call infallible.

**Comment by hhromic** [comment]
But yes, if it fails, it returns the empty object. Fully agree with what you commented later.

**Comment by hhromic** [comment]
> I wonder if we should prevent the err-assign operation to assign a value to the `ok` target _if_ that target already exists. It would be a breaking change, and it might be confusing to people now that I think about it, but it would at least solve the above issue.

I was asking not so long ago about your proposed error handling behaviour for Vector in Discord. The rationale was that Vector currently behaves pretty much like GoLang does, which is probably a good idea to maintain.

If a function fails, you should not expect anything valid/useful in the `ok` target. Changing this behaviour is powerful but agree can be very confusing.

From what I understand, the preferred programming pattern for this situation should be:
```coffee
parsed, err = parse_json(.message)
if err != null {
  log(err, level: "error")
  abort
}
. |= parsed
```

And, again as you said, you gain nothing really :)

**Comment by JeanMertz** [comment]
> I think you meant this?
> 
> ```coffee
> ok, err = parse_json(.message)
> # "function call error for \"parse_json\" at (10:30): expected \"string\", got \"null\""
> ```
> 
> You were using the `!` modifier, thus indeed making the call infallible.

No, I meant it as I wrote it. The `!` in `parse_json!` is there to catch the error at runtime when the field isn't a JSON-encoded string. After that, you can get a non-object back, which is why I proposed catching that with `ok, err |= ...`.

**Comment by hhromic** [comment]
> > I think you meant this?
> > ```coffee
> > ok, err = parse_json(.message)
> > # "function call error for \"parse_json\" at (10:30): expected \"string\", got \"null\""
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You were using the `!` modifier, thus indeed making the call infallible.
> 
> No, I meant it as I wrote it. The `!` in `parse_json!` is there to catch the error at runtime when the field isn't a JSON-encoded string. After that, you can get a non-object back, which is why I proposed catching that with `ok, err |= ...`.

Ah, I missed the point that what you wrote is _currently_ not possible but _would_ work in a potential solution 🤦 .
Sorry about that!

**Comment by nabokihms** [comment]
With vector 0.22, users will be able to check the type first and then parse. It should be more efficient than `. |= object!(parse_json!(.message))` if there mostly negative cases on the hot path.

```lua
if is_json(.message, variant: "object") {
    res, err = parse_json(.message)
    ...
}
```

**Comment by nabokihms** [comment]
@JeanMertz, we can apply the same logic as we did for the is_json function but slightly rearrange the order
1. Check the first non-whitespace byte according to these arms https://github.com/vectordotdev/vector/blob/b530718e961b6d4a0a4ba9428fdcccb41a8a8e18/lib/vrl/stdlib/src/is_json.rs#L20-L30
2. If provided variant and the byte match - try to parse the message.

It should be performant enough for all JSON-formatted messages without an enormous number of spaces at the start.

Example:
```lua
parse_json(.message, variant: "object")
```
